src/library/dao/playlistdao.cpp

```diff
@@ -980,18 +982,16 @@ void PlaylistDAO::shuffleTracks(const int playlistId, const QList<int>& position
         trackPositionIds.insert(trackBPosition, trackAId);
         newPositions.swap(newPositions.indexOf(trackAPosition),
                           newPositions.indexOf(trackBPosition));
-        QString swapQuery = "UPDATE PlaylistTracks SET position=%1 "
-                "WHERE position=%2 AND playlist_id=%3";
-        query.exec(swapQuery.arg(QString::number(-1),
-                                 QString::number(trackAPosition),
-                                 QString::number(playlistId)));
-        query.exec(swapQuery.arg(QString::number(trackAPosition),
-                                 QString::number(trackBPosition),
-                                 QString::number(playlistId)));
-        query.exec(swapQuery.arg(QString::number(trackBPosition),
-                                 QString::number(-1),
-                                 QString::number(playlistId)));
 
+        qSwap(playListTrackIds[trackAPosition - 1],
+              playListTrackIds[trackBPosition - 1]);
```

Comment:
Yes, the last issue can be solved if you iterate only though the "positions" and "newPositions", using your original list as lookup. A Hash would be required only if we allow a non consecutive numbering. 

You should place an DEBUG_ASSERT for this.